modules/pricer/src/main/java/com/opengamma/strata/pricer/curve/SyntheticRatesCurveCalibrator.java

```diff
@@ -158,37 +160,44 @@ public ImmutableMarketData marketData(
     for (Index idx : Sets.intersection(inputProvider.getTimeSeriesIndices(), indicesRequired)) {
       ts.put(IndexQuoteId.of(idx), inputProvider.timeSeries(idx));
     }
-
     LocalDate valuationDate = inputProvider.getValuationDate();
     ImmutableList<CurveDefinition> curveGroups = group.getCurveDefinitions();
     // Create fake market quotes of 0, only to be able to generate trades
-    Map<MarketDataId<?>, Double> mapId0 = new HashMap<>();
+    Map<MarketDataId<?>, Object> mapId0 = new HashMap<>();
     for (CurveDefinition entry : curveGroups) {
       ImmutableList<CurveNode> nodes = entry.getNodes();
       for (int i = 0; i < nodes.size(); i++) {
         for (MarketDataId<?> key : nodes.get(i).requirements()) {
-          mapId0.put(key, 0.0d);
+          ArgChecker.isTrue((key instanceof QuoteId) || (key instanceof FxRateId),
+              "For Synthetic curve calibration, the MarketDataId must be QuoteId or FxRateId");
+          if (key instanceof QuoteId) {
+            mapId0.put(key, 0.0d);
+          }
         }
       }
     }
-    ImmutableMarketData marketQuotes0 = ImmutableMarketData.of(valuationDate, mapId0);
     // Generate market quotes from the trades
     Map<MarketDataId<?>, Object> mapIdSy = new HashMap<>();
-    for (CurveDefinition entry : curveGroups) {
-      ImmutableList<CurveNode> nodes = entry.getNodes();
-      for (CurveNode node : nodes) {
-        ResolvedTrade trade = node.resolvedTrade(1d, marketQuotes0, refData);
-        double mq = measures.value(trade, inputProvider);
-        MarketDataId<?> k = node.requirements().iterator().next();
-        mapIdSy.put(k, mq);
-      }
-    }
     // Generate quotes for FX pairs. The first currency is arbitrarily selected as starting point. 
     // The crosses are automatically generated by the MarketDataFxRateProvider used in calibration.
     for (int loopccy = 1; loopccy < ccyRequired.size(); loopccy++) {
       CurrencyPair ccyPair = CurrencyPair.of(ccyRequired.get(0), ccyRequired.get(loopccy));
       FxRateId fxId = FxRateId.of(ccyPair);
       mapIdSy.put(fxId, FxRate.of(ccyPair, inputProvider.fxRate(ccyPair)));
+      mapId0.put(fxId, FxRate.of(ccyPair, inputProvider.fxRate(ccyPair)));  // Required to generate the trade
```

Comment:
Because `MarketData` created this map is used in `CurveNode` for cross currency products without `MarketDataFxRateProvider`, FX crosses are not generated. Thus the exact `FxRateId` is required.